### ISSUE TITLE
Packaging firefox extensions

### ISSUE THREAD

**Issue by romildo** [NixOS/nixpkgs#15959]
I want to package the [Arc Firefox Theme](https://github.com/horst3180/arc-firefox-theme), which complements the [Arc Theme](https://github.com/horst3180/arc-theme) for desktop environments based on GTK. It is a set of three extensions for Firefox (`*.xpi` files) that should be installed into the Firefox extensions directory (`${firefox-unwrapped}/lib/firefox-xx.x.x/browser/extensions/` I think).

Since the store is read-only, that can not be done directly. Therefore a means of installing Firefox extensions should be created in NixOS.

Would it be feasible to have a directory `/run/current-system/sw/share/firefox/extensions` containing symbolic links to installed extensions (`*.xpi` files), and make `${firefox-unwrapped}/lib/firefox-xx.x.x/browser/extensions` be a symbolic link to it?

**Comment by joachifm** [comment]
To elaborate, the idea is that you create a suitable directory hierarchy, copying in extension data + firefox as necessary. I also wanted to be able to put in `prefs.js`, too, for a fully customized deployment.

**Comment by romildo** [comment]
cc @edolstra I am not experienced enough to implement it. Probably Firefox maintainers will need to collaborate on it.

**Comment by edolstra** [comment]
Making `${firefox-unwrapped}/lib/firefox-xx.x.x/browser/extensions` a symlink is not feasible because it wouldn't work if Firefox is installed via `nix-env`.

Maybe Firefox has an environment variable similar to `MOZ_PLUGIN_PATH` for extensions?

**Comment by joachifm** [comment]
Relevant docs
https://developer.mozilla.org/en-US/Firefox/Enterprise_deployment
https://developer.mozilla.org/en-US/Add-ons/Installing_extensions
https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Customizing_Firefox#Including_extensions_with_your_distribution_of_Firefox

**Comment by joachifm** [comment]
A crude variant of this is at https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/browsers/tor-browser-bundle/default.nix#L159

**Comment by oxij** [comment]
Since this is an item on both SLNOS TODO and my own TODO I have a couple
of comments.

First, you have to have `firefoxWithPackages` for it to be useful (e.g.
for running both tor-browser and firefox).

Second, `tor-browser-bundle` expression indeed has a lot of hints, but
it cheats by just placing those xpis into the same output with the
browser. Meanwhile `firefoxWithPackages` needs to be able to run the
browser from one /nix/store place and to load xpis from another (unless
you're ok with recompiling `firefox` every time). It will probably just
work with some `buildEnv` hackery, but it might need some patching of
`firefox` (to make it consult env or something) since tor-browser
patches that part of firefox extensively ("bundle").

Anyway, SLNOS overlords welcome your efforts. It's also a high-priority
TODO item on my own list, so I will do it eventually (but I don't have
time for any serious nix work ATM), so feel free to ask for any help if
you wish to implement it.

I would start by packaging a single xpi and experimenting with firefox
to make it load it from /nix/store (with prefs, by merging them into a
single buildEnv, etc).

**Comment by joachifm** [comment]
It doesn't have to recompile anything when you change the xpis so I don't consider it cheating.

**Comment by joachifm** [comment]
That is, I declare both the prefs and extensions via config.nix and change them at will with minimal rebuilding.

**Comment by joachifm** [comment]
(the current version doesn't put the xpis in the same output, but they are symlinked).

**Comment by oxij** [comment]
> It doesn't have to recompile anything when you change the xpis so I don't consider it cheating.

It repackages binaries made to be placed anywhere ("bundle").

> That is, I declare both the prefs and extensions via config.nix and change them at will with minimal rebuilding.

Can you show an example of that, please?

**Comment by joachifm** [comment]
>  It repackages binaries made to be placed anywhere ("bundle").

I don't consider repackaging the runtime to be recompiling firefox. I'll grant that re-packaging an existing firefox runtime is inefficient, compared to patching, but patching adds a lot of friction on my end and so is infeasible. Naive application of `buildEnv` does not work.

> Can you show an example of that, please?

Sure (this is with unpublished changes ontop of master):
```nix
{
  allowUnfree = true;
  allowBroken = false;

  packageOverrides = super: let self = {
    firefoxExtensions.treestyletab = super.stdenvNoCC.mkDerivation rec {
      name = "treestyletab-${version}";
      version = "0.19.2017031101";
      src = super.fetchgit { # needs submodules for makexpi, todo: fetch
        url = "https://github.com/piroor/treestyletab";
        rev = "refs/tags/${version}";
        sha256 = "01wjb0r7dr14xg2g3g588fx436nxp6myyik2ymkag81gv1v0fia5";
      };
      nativeBuildInputs = with super; [ zip ];
      buildPhase = ''
        $shell makexpi/makexpi.sh -n treestyletab -o
      '';
      installPhase = ''
        install -m 444 -D treestyletab.xpi $out/treestyletab@piro.sakura.ne.jp.xpi
      '';
    };

    tor-browser-bundle = super.tor-browser-bundle.override {
      audioSupport = true;
      mediaSupport = true;

      extraExtensions = with self.firefoxExtensions; [ treestyletab ];
      extraPrefs = ''
        lockPref("extensions.torbutton.security_slider", 1); // 4=compat (default), 2=mid, 1=high

        // Work-around flickering in treestyletab
        pref("browser.tabs.animate", false);

        // For extraExtensions
        pref("xpinstall.signatures.required", false);
      '';
    };
  }; in self;
}
```

Changing prefs or extensions gives a nix-rebuild time of
```
nix-build . -A tor-browser-bundle  0.55s user 0.05s system 15% cpu 3.961 total
```
most of which is spent in the post fixup phase. For my purposes this feature is working to the point that the exact mechanics of it are irrelevant, though they could be greatly improved, for sure.

**Comment by joachifm** [comment]
Note also that the old binary bundle runs entirely from the Nix store (as far as the "bundled" payload is concerned), it only copies the initial bookmarks file into the runtime profile directory on first run, My main motivation for the for-source version is not firefox-with-extensions but tor-browser-bundle-from-source.

**Comment by joachifm** [comment]
... but eventually I'd hope that it could be subsumed by a more general construct.

**Comment by leo60228** [comment]
@joachifm Would/could this be merged into master?

**Comment by joachifm** [comment]
@leo60228 the pasted example above works as-is now. Due to lack of interest I've had little motivation to pursue it further.

**Comment by leo60228** [comment]
still important to me, home-manager's implementation could be in nixpkgs and seems pretty popular

**Comment by joelmccracken** [comment]
I would LOVE this feature. I actually spent a decent chunk of time trying to figure out how to install firefox extensions automatically a long time ago, and it was quite hard. But anyway I'd love to see this, maybe I could hack on it at some point.

**Comment by leo60228** [comment]
This is definitely still an issue. I have a custom implementation of this but it's pretty tied to home-manager and isn't really in a state to PR it.

**Comment by YoshiRulz** [comment]
I believe this was resolved by #91724.

**Comment by samueldr** [comment]
Assumed fixed.
